cranelift/codegen/src/isa/zkasm/inst/emit.rs

```diff
@@ -912,61 +912,83 @@ impl MachInstEmit for Inst {
                 let base = from.get_base_register();
                 let offset = from.get_offset_with_state(state);
                 let rd = allocs.next_writable(rd);
-                let insn = match from {
-                    AMode::RegOffset(r, off, _) => format!(
-                        "$ => {} :MLOAD({})\n",
-                        reg_name(rd.to_reg()),
-                        access_reg_with_offset(r, off),
-                    ),
+                match from {
+                    AMode::RegOffset(r, off, _) => {
+                        // For now we only support zero offsets.
+                        assert_eq!(off, 0);
+                        if r == context_reg() {
+                            put_string(&format!("0 => {}\n", reg_name(rd.to_reg())), sink);
+                        } else {
+                            // TODO(akashin): Get rid of this unsound logic when 32-bit registers
+                            // are stored without shifts.
+                            put_string(&format!("${{ {} >> 32 }} => E\n", reg_name(r)), sink);
```

Comment:
I think migrating to masking can take a time, so maybe it worth to put ARITH here for now? Also, it is important to keep both versions of 32-bit implementation correct, if we will want to compare this two versions (I believe that masking can be better, but it will be good to compare and be shure)

Reply:
True, I went through a few iterations here and tried using AirthDiv64/32 and Shru64/32 but wasn't able to get that version to work, either due to the lack of registers or due to failure in the assert at the end of the program that I wasn't able to figure out.

So I decided to put a hack here and instead try to get the full SHA256 working end-to-end first and then return and fix this part (either with ARITH-based approach or with masking).

Regarding comparison, my thinking was that this version gives us a lower bound on the performance of the version with shifts (i.e. true version with shifts will be slower), so if we improve on that lower bound, that's a good enough signal to switch.

Reply:
No argument that this biases comparison in favour of current 32bit strategy. So we'll have to have that in mind.

I'm fine to try other things here, but if we submit this now, this will unblock work on globals and data segments which can be done in parallel with fixing this part.